packages/firestore/src/local/local_serializer.ts

```diff
@@ -231,6 +236,81 @@ export class LocalSerializer {
       queryProto
     );
   }
+
+  /** Encodes a DbBundle to a Bundle. */
+  fromDbBundle(dbBundle: DbBundle): Bundle {
+    return {
+      id: dbBundle.bundleId,
+      createTime: this.fromDbTimestamp(dbBundle.createTime),
+      version: dbBundle.version
+    };
+  }
+
+  /** Encodes a BundleMetadata to a DbBundle. */
+  toDbBundle(metadata: bundleProto.BundleMetadata): DbBundle {
+    return {
+      bundleId: metadata.id!,
+      createTime: this.toDbTimestamp(
+        this.remoteSerializer.fromVersion(metadata.createTime!)
+      ),
+      version: metadata.version!
+    };
+  }
+
+  /** Encodes a DbNamedQuery to a NamedQuery. */
+  fromDbNamedQuery(dbNamedQuery: DbNamedQuery): NamedQuery {
+    return {
+      name: dbNamedQuery.name,
+      query: this.fromBundledQuery(dbNamedQuery.bundledQuery),
+      readTime: this.fromDbTimestamp(dbNamedQuery.readTime)
+    };
+  }
+
+  /** Encodes a NamedQuery from bundle proto to a DbNamedQuery. */
+  toDbNamedQuery(query: bundleProto.NamedQuery): DbNamedQuery {
+    return {
+      name: query.name!,
+      readTime: this.toDbTimestamp(
+        this.remoteSerializer.fromVersion(query.readTime!)
+      ),
+      bundledQuery: query.bundledQuery!
+    };
+  }
+
+  /**
+   * Encodes a `BundledQuery` from bundle proto to a Query object.
+   *
+   * This reconstructs the original query used to build the bundle being loaded,
+   * including features exists only in SDKs (for example: limit-to-last).
+   */
+  fromBundledQuery(bundledQuery: bundleProto.BundledQuery): Query {
+    const query = this.remoteSerializer.convertQueryTargetToQuery({
+      parent: bundledQuery.parent!,
+      structuredQuery: bundledQuery.structuredQuery!
+    });
+    if (bundledQuery.limitType === 'LAST') {
+      return query.withLimitToLast(query.limit);
+    }
+    return query;
+  }
+
+  /** Encodes a NamedQuery proto object to a NamedQuery model object. */
+  fromProtoNamedQuery(namedQuery: bundleProto.NamedQuery): NamedQuery {
+    return {
+      name: namedQuery.name!,
+      query: this.fromBundledQuery(namedQuery.bundledQuery!),
+      readTime: this.remoteSerializer.fromVersion(namedQuery.readTime!)
+    };
+  }
+
+  /** Encodes a BundleMetadata proto object to a Bundle model object. */
+  fromBundleMetadata(metadata: bundleProto.BundleMetadata): Bundle {
+    return {
+      id: metadata.id!,
+      version: metadata.version!,
+      createTime: this.remoteSerializer.fromVersion(metadata.createTime!)
+    };
+  }
```

Comment:
Ideally, we should not add any more code to the main SDK for "optional" features. Do you mind exploring what it would entail to make these functions free-standing? They would then look like:


function fromBundleMetadata(serializer: RemoteSerializer, metadata: bundleProto.BundleMetadata): Bundle {
    return {
      id: metadata.id!,
      version: metadata.version!,
      createTime: serializer.fromVersion(metadata.createTime!)
    };
  }
```

If you have to make a ton of state public, you can look at what we did here: https://github.com/firebase/firebase-js-sdk/blob/5cf39ab8cacd5f055d609acf79a859beda1032d3/packages/firestore/src/remote/datastore.ts#L51

DatastoreImp has public state, but in itself is private.